Adjustment for ODT redesign branch and minor improvements:

LarsGit223 opened the pull request:
- improved support for languages/text direction (ltr or rtl)
- render boxes as tables (so we lose round corners but get formating inside of the boxes)
- Paragraph and Span styles will be editable in LibreOffice after export

The Wrap plugin will be backward compatible to previous ODT plugin versions but new features will only work with the redesign branch of the ODT plugin.

With the current release of the ODT plugin everything should be as it was before except support for languages: text direction and alignment should now be set according to the language.

selfthinker commented:
Sorry for the really long wait. I've just tried to look at this but whatever I do, I cannot open file that gets created from Wrap's example.txt because I always get an error from LibreOffice Writer when I try to open it:

> Read Error.
> Format error discovered in the file in sub-document content.xml at 483,89(row,col).

I first thought this must be due to my old LibreOffice version but an update to the latest version didn't help. (I'm using the current master of DokuWiki. I also tried using the latest stable version but had other issues unrelated to ODT.)
I get the same without your changes, but I need it to work in order to test them.

LarsGit223 commented:
Well, I really have to re-test this PR as I did not use it for a long time. Maybe changes in-between on the ODT plugin broke something. I will get back with the results.

LarsGit223 commented:
Hmmm...works on my installation.

I took the latest stable ODT release and my master fork of the wrap plugin. On some development version of DokuWiki (shouldn't matter). I also copied example.txt to a fresh wiki page to make sure it hasn't changed in the meantime. I switched back from media selector ```screen``` to ```print``` in my ODT config settings. No problems so far (unfortunately).

Some questions:
- Which ODT release are you using? If I understood you right you are not using the latest stable version due to some other problems.
- Can you post a screenshot of your ODT config settings?

selfthinker commented:
Thanks for checking.
I used the latest version of LibreOffice, the current master branch of the ODT plugin, the current master branch of DokuWiki and this branch of the Wrap plugin but also the master branch.
I think I only changed one single ODT setting (`media_sel`) but also tried without that. I can double check tomorrow.

I have a vague memory that I had an issue like this before, but no idea what I did.
I just got a new laptop and will re-test everything when I've set things up on the weekend.

selfthinker commented:
Ah, thanks, got it:

> XML Parsing Error: mismatched tag. Expected: </table:table-row>.
> Location: file:///[path to file]/content.xml
> Line Number 365, Column 90:
> (If you specify a language not listed there, it simply won&#39;t do anything.)</text:p></office:text></office:body></office:document-content>
> -----------------------------------------------------------------------------------------^

LarsGit223 commented:
Ok, it looks like the document reaches it's end and a table was not finished. The wrap boxes are rendered as tables. But it's strange that I do not have the problem. Could you post me the complete file please?

selfthinker commented:
Here is the example output: [wrap-example.zip](https://github.com/selfthinker/dokuwiki_plugin_wrap/files/1777490/wrap-example.zip).

Also, for completeness' sake: I'm running PHP version 7.1.1 on Windows 10. I'm not sure how likely it is that that could be the cause.

LarsGit223 commented:
I updated to PHP version 7.1.15, no error  :worried:

I had a look at the source for ODT tables and could not find anything that looks like a definitive error to me. But I did find an unclean thing in a method signature. Not sure if it makes a difference because I still get no error/broken document. But I adjusted the code in branch ```selfthinker```. It would be great if you could test it. Except the mentioned change, the branch is equal to ```master```.

selfthinker commented:
I just tested that branch and still get the very same problem (I also double checked `content.xml`).

I then tried to find what is the most reduced case which produces the error and found this already triggers it (and the rest of the page is fine without it):
```
<WRAP column>
Column
</WRAP>
```

Can you think of anything else which could be different about our setups?

LarsGit223 commented:
Good news. With a wiki page only having the content
```
<WRAP third column>...content...</WRAP>
```
I can re-produce the problem. But I am not getting the broken document if I export the complete Wrap plugin example page.

LarsGit223 commented:
Ahh...no. Not an error. For the ODT export there MUST be a group surrounding the columns. But if I use
```
<WRAP group>
<WRAP third column>...content...</WRAP>
</WRAP>
```
then the exported document is valid.

LarsGit223 commented:
Could you please post the minimalistic broken ODT document so that I could have a look please? Thanks in advance.

> Can you think of anything else which could be different about our setups?

I already installed PHP 7.1 and the latest official version of DokuWiki. Wrap and ODT plugin are now the official versions (except the little change for the attempted bugfix). Maybe the plugin config. But I played around with it, no difference.

LarsGit223 commented:
My ```local.php``` includes the following config settings for ODT:
``` php
$conf['plugin']['odt']['showpdfexportbutton'] = 1;
$conf['plugin']['odt']['index_in_browser'] = 'display placeholder';
$conf['plugin']['odt']['outline_list_style'] = 'Numbers';
$conf['plugin']['odt']['olist_label_align'] = 'left';
$conf['plugin']['odt']['usestyles'] = 'wrap';
$conf['plugin']['odt']['import_basic'] = 'Yes';
$conf['plugin']['odt']['toc_styleL2'] = 'color:red;';
$conf['plugin']['odt']['toc_styleL3'] = 'color:blue;font-weight:bold;';
```
Could you please post yours just to compare it once? Thanks.

selfthinker commented:
I've changed my reduced test case to be wrapped in a `group` wrap but I get the same results: 
[wrap-reduced-test.zip](https://github.com/selfthinker/dokuwiki_plugin_wrap/files/1823115/wrap-reduced-test.zip)

I didn't have any ODT config settings set. (I played around with two earlier.)
I have now copied your config and the output was still broken.

I've also switched to DokuWiki's `stable` branch just because Wrap is currently broken in other ways when using `master` (due to the [section edit changes](https://github.com/splitbrain/dokuwiki/pull/2220/files#r175291349)). I wanted to be sure the other bugs are not interfering with this one (although the file is corrupt in both branches).
And I've disabled all other plugins (except ODT and Wrap).

LarsGit223 commented:
Thanks for providing the broken document. As before the table is simply not closed:
``` xml
<table:table table:style-name="PluginODTAutoStyle_Table_1">
    <ColumnsPlaceholder1>
        <table:table-row table:style-name="PluginODTAutoStyle_TableRow_2">
            <table:table-cell office:value-type="string" table:style-name="PluginODTAutoStyle_TableCell_3">
                <text:p text:style-name="PluginODTAutoStyle_Paragraph_4">Column</text:p>
            </table:table-cell>

</office:text>
</office:body>
</office:document-content>
```
That means the ```<WRAP group>``` and ```<WRAP column>``` and the closing ```</WRAP>``` for the column have been processed properly because at least the ```table:table-cell``` element is closed. The row and the table itself should have been closed on the closing ```</WRAP>``` for the group but the broken document looks as if it wouldn't have been called. I assumed a not initialized variable causing such random effects but did not find one.

I did find a wrongly referenced static class variable in the Wrap helper. In ```wrap/helper.php``` there is this variable defined at the begining:
``` php
    static protected $column_count = 0;
```
This is wrongly used as
``` php
    $this->column_count
```
Could you try to replace all occurrences of ```$this->column_count``` with ```self::$column_count``` (only 4)? After that re-test the export.

selfthinker commented:
Thanks. I've replaced those and re-tested, but unfortunately the file is still broken. :(

LarsGit223 commented:
Thanks a lot again for testing. I think there is no other way to get some progress than to put some debug outputs in it. I will prepare a branch with debug outputs and will let you know when it's ready for testing.

LarsGit223 commented:
I have put debug outputs into the ```selfthinker```ODT branch. When you are going to test it, make sure to enable the ```allowdebug``` config option first. Then it should generate debug outputs in the following form:
```
18:18:45 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableOpenUseProperties
18:18:45 127.0.0.1: >>> ODTDocument::tableOpenUseProperties
18:18:45 127.0.0.1: <<< ODTDocument::tableOpenUseProperties
18:18:45 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableOpenUseProperties
```
I hope this will give me some insight.

selfthinker commented:
Thanks. This is the output:

```
22:48:20 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableOpenUseProperties
22:48:20 127.0.0.1: >>> ODTDocument::tableOpenUseProperties
22:48:20 127.0.0.1: <<< ODTDocument::tableOpenUseProperties
22:48:20 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableOpenUseProperties
22:48:20 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableRowOpenUseProperties
22:48:20 127.0.0.1: >>> ODTDocument::tableRowOpenUseProperties
22:48:20 127.0.0.1: <<< ODTDocument::tableRowOpenUseProperties
22:48:20 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableRowOpenUseProperties
22:48:20 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableAddColumnUseProperties
22:48:20 127.0.0.1: >>> ODTDocument::tableAddColumnUseProperties
22:48:20 127.0.0.1: <<< ODTDocument::tableAddColumnUseProperties
22:48:20 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableAddColumnUseProperties
22:48:20 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableCellOpenUseProperties
22:48:20 127.0.0.1: >>> ODTDocument::tableCellOpenUseProperties
22:48:20 127.0.0.1: <<< ODTDocument::tableCellOpenUseProperties
22:48:20 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableCellOpenUseProperties
22:48:20 127.0.0.1: >>> renderer_plugin_odt_page::tablecell_close
22:48:20 127.0.0.1: >>> ODTDocument::tableCellClose
22:48:20 127.0.0.1: <<< ODTDocument::tableCellClose
22:48:20 127.0.0.1: <<< renderer_plugin_odt_page::tablecell_close
```

LarsGit223 commented:
Thanks for testing again. That really looks like the functions for closing the row and table are not called. I expected the problem to be in the ODT plugin itself but it looks like it is in the ODT helper of the wrap plugin. I will prepare a version with debug outputs for it also and notify you once it's ready.

LarsGit223 commented:
I have put some debug outputs into the wrap plugin, see branch ```debug``` in my repository (https://github.com/LarsGit223/dokuwiki_plugin_wrap/tree/debug).

selfthinker commented:
Thanks. This is the debug output now (with odt debugging still in place):

```
22:46:25 127.0.0.1: >>> helper_plugin_wrap::renderODTElementOpen wrap_group
22:46:25 127.0.0.1: <<< helper_plugin_wrap::renderODTElementOpen other
22:46:25 127.0.0.1: >>> helper_plugin_wrap::renderODTElementOpen wrap_column
22:46:25 127.0.0.1: >>> helper_plugin_wrap::renderODTOpenColumn
22:46:25 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableOpenUseProperties
22:46:25 127.0.0.1: >>> ODTDocument::tableOpenUseProperties
22:46:25 127.0.0.1: <<< ODTDocument::tableOpenUseProperties
22:46:25 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableOpenUseProperties
22:46:25 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableRowOpenUseProperties
22:46:25 127.0.0.1: >>> ODTDocument::tableRowOpenUseProperties
22:46:25 127.0.0.1: <<< ODTDocument::tableRowOpenUseProperties
22:46:25 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableRowOpenUseProperties
22:46:25 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableAddColumnUseProperties
22:46:25 127.0.0.1: >>> ODTDocument::tableAddColumnUseProperties
22:46:25 127.0.0.1: <<< ODTDocument::tableAddColumnUseProperties
22:46:25 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableAddColumnUseProperties
22:46:25 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableCellOpenUseProperties
22:46:25 127.0.0.1: >>> ODTDocument::tableCellOpenUseProperties
22:46:25 127.0.0.1: <<< ODTDocument::tableCellOpenUseProperties
22:46:25 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableCellOpenUseProperties
22:46:25 127.0.0.1: <<< helper_plugin_wrap::renderODTOpenColumn
22:46:25 127.0.0.1: <<< helper_plugin_wrap::renderODTElementOpen column
22:46:25 127.0.0.1: >>> helper_plugin_wrap::renderODTElementClose column
22:46:25 127.0.0.1: >>> helper_plugin_wrap::renderODTCloseColumn
22:46:25 127.0.0.1: >>> renderer_plugin_odt_page::tablecell_close
22:46:25 127.0.0.1: >>> ODTDocument::tableCellClose
22:46:25 127.0.0.1: <<< ODTDocument::tableCellClose
22:46:25 127.0.0.1: <<< renderer_plugin_odt_page::tablecell_close
22:46:25 127.0.0.1: <<< helper_plugin_wrap::renderODTCloseColumn
22:46:25 127.0.0.1: <<< helper_plugin_wrap::renderODTElementClose column
```

selfthinker commented:
I'm not sure if #163 could be related? I get broken HTML with DokuWiki's master branch but not stable.
But I get the broken ODT export also with the stable branch. So, that might not be relevant.

LarsGit223 commented:
I've read #163. It doesn't sound like a problem that is happening in a coincidental way. So right now I would say it is not related.

LarsGit223 commented:
Here is my output from my working scenario based on _2017-02-19e "Frusterick Manners"_. Please do not wonder about the longer output. I got more debug outputs in the ODT plugin because I troubleshooted and fixed a different problem. So here is the output:
```
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::p_open
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::p_open
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::cdata 
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::p_close
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::p_close
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTElementOpen group
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTElementOpen group
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTElementOpen wrap_column
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTOpenColumn
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableOpenUseProperties
10:30:27 127.0.0.1: >>> ODTDocument::tableOpenUseProperties
10:30:27 127.0.0.1: <<< ODTDocument::tableOpenUseProperties
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableOpenUseProperties
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableRowOpenUseProperties
10:30:27 127.0.0.1: >>> ODTDocument::tableRowOpenUseProperties
10:30:27 127.0.0.1: <<< ODTDocument::tableRowOpenUseProperties
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableRowOpenUseProperties
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableAddColumnUseProperties
10:30:27 127.0.0.1: >>> ODTDocument::tableAddColumnUseProperties
10:30:27 127.0.0.1: <<< ODTDocument::tableAddColumnUseProperties
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableAddColumnUseProperties
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableCellOpenUseProperties
10:30:27 127.0.0.1: >>> ODTDocument::tableCellOpenUseProperties
10:30:27 127.0.0.1: <<< ODTDocument::tableCellOpenUseProperties
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableCellOpenUseProperties
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTOpenColumn
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTElementOpen column
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::p_open
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::p_open
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::cdata contentdsfgsdfg23
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::p_close
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::p_close
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTElementClose column
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTCloseColumn
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::tablecell_close
10:30:27 127.0.0.1: >>> ODTDocument::tableCellClose
10:30:27 127.0.0.1: <<< ODTDocument::tableCellClose
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::tablecell_close
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTCloseColumn
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTElementClose column
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTElementOpen wrap_column
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTOpenColumn
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableAddColumnUseProperties
10:30:27 127.0.0.1: >>> ODTDocument::tableAddColumnUseProperties
10:30:27 127.0.0.1: <<< ODTDocument::tableAddColumnUseProperties
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableAddColumnUseProperties
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableCellOpenUseProperties
10:30:27 127.0.0.1: >>> ODTDocument::tableCellOpenUseProperties
10:30:27 127.0.0.1: <<< ODTDocument::tableCellOpenUseProperties
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableCellOpenUseProperties
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTOpenColumn
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTElementOpen column
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::p_open
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::p_open
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::cdata content23
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::p_close
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::p_close
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTElementClose column
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTCloseColumn
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::tablecell_close
10:30:27 127.0.0.1: >>> ODTDocument::tableCellClose
10:30:27 127.0.0.1: <<< ODTDocument::tableCellClose
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::tablecell_close
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTCloseColumn
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTElementClose column
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTElementClose group
10:30:27 127.0.0.1: >>> helper_plugin_wrap::renderODTCloseGroup count:2
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::tablerow_close
10:30:27 127.0.0.1: >>> ODTDocument::tableRowClose
10:30:27 127.0.0.1: <<< ODTDocument::tableRowClose
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::tablerow_close
10:30:27 127.0.0.1: >>> renderer_plugin_odt_page::_odtTableClose
10:30:27 127.0.0.1: >>> ODTDocument::tableClose
10:30:27 127.0.0.1: <<< ODTDocument::tableClose
10:30:27 127.0.0.1: <<< renderer_plugin_odt_page::_odtTableClose
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTCloseGroup
10:30:27 127.0.0.1: <<< helper_plugin_wrap::renderODTElementClose group
```
I can say that the group is not opened and closed properly in your case. To be more precise is it opened but not closed. Also, in your case on opening of the group it is given as ```wrap_group``` and in my case as ```group```.

Ahhh, just noticed there is a configuration option ```noprefix``` for wrap. I think I never touched it. In my config it is set to ```tabs, group```.

The ODT helper doesn't seem to recognize the option properly in your case/with your settings. Anyway, the close is missing so it does not seem to be the root of the problem.

I have to adjust my previous post. I think it might be related to #163. I will have a further look.

LarsGit223 commented:
I think I was not precise in my previous post:
- because the ODT plugin right now does not recognize a group if ```wrap_group``` is passed, then the ODT document will always be broken
- if my config is used and ```group``` is passed then I assume the behaviour would be consistent to the broken HTML issue

selfthinker commented:
Argh, I looked up if I had any Wrap setting changed and my `noprefix` was missing `group`. Using only default settings (which put `group` back in) indeed fixed the whole thing, my ODT export file can now be opened and looks fine!
I'm very sorry for the additional work that has caused you and thankful that you took the time. I haven't worked on DokuWiki for a couple of years and just assumed my old `conf/local.php` was in a good state, but I must have experimented with something when I last worked on it and couldn't remember. I double checked all my ODT settings but none of my Wrap settings.

That didn't effect #163, so it will be unrelated.

I will now look at your actual pull request again. :)

LarsGit223 commented:
> Using only default settings (which put group back in) indeed fixed the whole thing, my ODT export file can now be opened and looks fine!

Great. But I would still judge this as a bug in the ODT helper and I think I will provide a PR soon, which will make sure it works with and without having ```group``` in ```noprefix```.

> I'm very sorry for the additional work that has caused you and thankful that you took the time.

Never mind. I guess you had a lot of additional work yourself. Thanks for all the testing.

> I will now look at your actual pull request again. :)

Nice. Please be aware of my comment in the commit message of the second commit in this PR:

> This means floating works well if there are just one or two frames on the same y position/height.
> What does not work well is the case that there are too much frames and they do not fit in on the
> same 'line' so that some would need to be wrapped below the other frames. In that case frames will overlap.
> 
> This makes the wrap example page look quite ugly regarding frames.
> But it works well if you have got just a text paragraph and e.g. one frame to include a picture
> or table which shall be surrounded by the text of the paragraph.
> 
So don't be shocked about the bad positioning of boxes if you export the wrap example page.

selfthinker commented:
> Great. But I would still judge this as a bug in the ODT helper and I think I will provide a PR soon, which will make sure it works with and without having `group` in `noprefix`.

Yes, that would be good. Someone could easily not understand what `group` is used for and delete it without knowing the consequences.

selfthinker commented:
I have now compared the output of the syntax example page from two different sets of things:
* DokuWiki's stable, ODT's publish and Wrap's slightly changed master branch to get the output most people would get right now
* DokuWiki's master, ODT's master and this branch of the Wrap plugin (but with section editing stuff removed so that there are no other errors due to #163) to get the output of these changes

(Because of your comment about the boxes I have also tested another simple page with a few boxes and text in between.)

What has visually **improved** between these two versions:
* text direction (as you mentioned in the description and commit message)
* lo (less significant) had a bigger font size previously and has now the same size as the other standard text
* code blocks have no boxes around them anymore; that's probably not a Wrap change but a general change in the ODT plugin, but I think it's an improvement in comparison
* multi columns look evened out, but that might just be a coincidence

What got **worse**:
* the 3 columns at the top are indented; is that intentional? I think it looked better before

Boxes and notes are pretty much **broken** in the syntax example page, but they are also broken in the current stable versions and look okay when there is just one box. So, that's probably okay as it is.

LarsGit223 commented:
> Boxes and notes are pretty much broken in the syntax example page, but they are also broken in the current stable versions and look okay when there is just one box. So, that's probably okay as it is.

Yes. The problem is that the ODT plugin is not really doing a rendering. That means it does not calculate the actual size a box would need in for width and height. This would be very complicated I guess (I had a look at cairo and pango but I think at the moment I will not work on improving this). Unfortunately there is no automatic way to encode the wanted behaviour in the ODT format. I can put tables in frames and give the frame a position relative to the current paragraph. But a frame is then not part of the paragraph and so I cannot place the next frame under the previous frame after floating/breaking into the next "line of frames".

> the 3 columns at the top are indented; is that intentional? I think it looked better before

No. I saved the exported example page and checked the table properties. It looks indented but in fact the problem is that the table has only 90% width instead of 100%. I will have a look.

LarsGit223 commented:
Ok, I checked the 3 columns and the behaviour is as implemented. But the layout is different than in the browser because in the exported ODT document columns are a real table. This is what happens:

Each of this columns has the tag ```<WRAP third column>```. Wrap third is defined as:
``` css
.dokuwiki .wrap_third {
    width: 30%;
    margin-right: 5%;
}
```
Now the problem is a change in the ODT plugin for better support of nested tables: it calculates the width of each column and the table width is assigned the sum of all columns width. In this case this makes up only 90%.

What is the effect of this CSS code in the browser? Have more space in-between the columns?

selfthinker commented:
Yes, that's for adding space in-between columns.

Ideally the table should add the 5% as a right padding inside the columns (except in the last column).
If that is not possible, could the table at least be moved to the left so that it doesn't look that odd? The table would still not be as narrow but at least vertically aligned.

LarsGit223 commented:
I gave the table 100% width, ignoring the width coming from the classes like ```third```. That way it does not look indented any more (it looked indented before because it had 90% width and was centered).

All columns will have the same right padding. At the moment the ODT plugin does not offer an API to modify the cell or column style after adding it. And I only know the last column on seeing the close of the surrounding group. But the table column was created and closed already to that time.

> Ideally the table should add the 5% as a right padding inside the columns (except in the last column).

So as written above this is not possible at the moment. And wouldn't the best solution be to have a right and left padding on the first and last column, and the columns in-between should have a left and right padding with half of that value?

selfthinker commented:
100% width for the columns table sounds good and looks good! :+1:
I'd be happy to merge. Unless you have anything to add?

While I was looking at some older PRs when ODT support was first implemented, I stumbled upon the [spreadsheet we originally used to track what is supported](https://docs.google.com/spreadsheets/d/1F4OE4Kc9TzV3JwXxJc62TgxSbkPmOuCNmfgyzWp3Xco/edit?usp=sharing) and what isn't . It might be worth updating that.
But I plan to add at least one new set of wrap classes before the next release.

LarsGit223 commented:
> I'd be happy to merge. Unless you have anything to add?

No, at least no code change. But maybe I should squash the commits. It's now 3 commits in this PR, each from a different year! Not nice for the commit history. Squashing it would make it look like one commit for all 3 changes.

selfthinker commented:
All 3 commits do different things and it makes sense to have them separately. That they are all from different years might look weird in the history and be impractical but I don't think that's a big deal. Unless you can change just the date without squashing?

LarsGit223 commented:
No, sorry. So if you want to keep the 3 commits then merge ahead please.

LarsGit223 commented:
:thumbsup: Cool, thanks for looking at it after all the time. I will post the PR regarding the noprefix issue in the comming week.
